src/rebar_cmd_prv.erl

```diff
@@ -49,36 +59,78 @@ format_error(Reason) ->
-    Port = erlang:open_port({spawn, Command}, [exit_status]),
-    shell_loop(Port, [], Timeout, {CmdName, Command}).
+do_internal({_CmdName, CmdDef}, _State) when
+    is_tuple(CmdDef) andalso
+        (erlang:tuple_size(CmdDef) < 2 orelse erlang:tuple_size(CmdDef) > 3)
+->
+    {error, "Invalid configuration: expected [{CmdName, Cmd, Opts}]."};
+do_internal({_CmdName, CmdDef}, State) when is_tuple(CmdDef) ->
+    {CmdName, Cmd, Opts} = normalize(CmdDef),
+    handle_cmd(exec(CmdName, Cmd, Opts), Opts, State);
+do_internal({undefined = _CmdName, false = _CommandFound}, _State) ->
+    {error, "No <command> (as in 'cmd <command>') given as input."};
+do_internal({CmdName, false = _CommandFound}, _State) ->
+    {error, "Command " ++ CmdName ++ " (as atom) is unknown. Check your configuration."}.
+
+find_command_in(undefined = _CmdName, _Config) ->
+    false;
+find_command_in(CmdName, Config) ->
+    lists:keyfind(list_to_atom(CmdName), 1, Config).
+
+normalize({CmdName, Cmd}) ->
+    normalize({CmdName, Cmd, []});
+normalize({CmdName, Cmd, Opts}) ->
+    {atom_to_list(CmdName), Cmd, Opts}.
 
-shell_loop(Port, Data, Timeout, {CmdName, Command}) ->
+exec(CmdName, Cmd, Opts) ->
+    Pid = self(),
+    Ref = erlang:make_ref(),
+    Timeout = get_opt(timeout, Opts, CmdName),
+    {MonitorPid, MonitorRef} = erlang:spawn_monitor(
+        fun() ->
+            not is_list(Cmd) andalso exit("call not expressed as a string"),
+            rebar_api:debug("Command ~s executing ~s with options ~p.", [CmdName, Cmd, Opts]),
+            Pid ! {Ref, os:cmd(Cmd)}
+        end
+    ),
     receive
-        {Port, {data, MoreData}} ->
-            io:format(MoreData),
-            shell_loop(Port, MoreData ++ Data, Timeout, {CmdName, Command});
-        {Port, {exit_status, 0}} ->
-            {ok, Data};
-        {Port, {exit_status, Error}} ->
-            Cmd = lists:flatten(Command),
-            {error,
-                "Command " ++
-                    CmdName ++
-                    " (" ++
-                    Cmd ++
-                    ") failed with exit status " ++
-                    integer_to_list(Error) ++ "."}
+        {Ref, Result} ->
+            {ok, {CmdName, Result}};
+        {'DOWN', MonitorRef, _Type, _Object, Info} ->
+            rebar_api:error("Command ~s died unexpectedly with ~p.", [CmdName, Info]),
+            {ok, {CmdName, died}}
     after Timeout ->
-        Cmd = lists:flatten(Command),
-        {error,
-            "Command " ++
-                CmdName ++
-                " (" ++
-                Cmd ++
-                ") timed out after " ++
-                integer_to_list(Timeout) ++ " ms."}
+        exit(MonitorPid, kill),
+        TimeOutS = integer_to_list(Timeout),
+        {error, "Command " ++ CmdName ++ " (" ++ Cmd ++ ") timed out after " ++ TimeOutS ++ " ms."}
     end.
+
+handle_cmd({error, Error}, _Opts, _State) ->
+    {error, Error};
+handle_cmd({ok, {_CmdName, died}}, _Opts, State) ->
+    {ok, State};
```

Comment:
should we move the error reporting to handle_cmd? seems like a better place

Reply:
Do you mean the `rebar_api:error` in `:exec/3`?

Reply:
20dd102